File: week2/task.js



// task1

const green_line_stations = ["Songshan","Nanjing Sanmin","Taipei Arena","Nanjing Fuxing","Songjiang Nanjing","Zhongshan","Beimen","Ximen","Xiaonanmen","Chiang Kai-Shek Memorial Hall","Guting","Taipower Building","Gongguan","Wanlong","Jingmei","Dapinglin","Qizhang","Xindian City Hall","Xindian","Xiaobitan"];

const messages={
    "Bob":"I'm at Ximen MRT station.",
    "Mary":"I have a drink near Jingmei MRT station.",
    "Copper":"I just saw a concert at Taipei Arena.",
    "Leslie":"I'm at home near Xiaobitan station.",
    "Vivian":"I'm at Xindian station waiting for you."
};

// let friends_location = ["","","Copper","","","","","Bob","","","","","","","Mary","","","","Vivian","Leslie"];

let friends_location=[];

for (const key in messages) {
    // if (Object.hasOwnProperty.call(messages, key)) {
    //     const element = messages[key];
    //     console.log(element);
    // }
    // console.log(messages[key]);
    let friend_idx = green_line_stations.map(e => messages[key].search(e)).findIndex(e => e>-1)
    friends_location[friend_idx] = key;
}
// console.log(friends_location);

let friends_index = [];
for (let index = 0; index < friends_location.length; index++) {
    if(friends_location[index]!== undefined){
        friends_index.push(index);
    }
}
console.log(friends_index);



function calulateStops(friend, current){
    // 確認存在
    // console.log("friend: "+friend);
    // console.log("current: "+current);

    let stop_count;

    // 朋友在小碧潭、自己不在小碧潭
    if(friend==19 & current!=19){
        // 計算從七張到自己所在地的間隔站數後+1
        stop_count = Math.abs(16 - current)+1;
    }
    // // 朋友在小碧潭、自己在小碧潭
    // else if(friend==19 & current==19){
    //     stop_count = 0;
    // }
    // 朋友不在小碧潭、自己在小碧潭
    else if(friend!=19 & current==19){
        stop_count = Math.abs(friend - 16)+1;
    }
    // 朋友不在小碧潭、自己不在小碧潭 OR 朋友在小碧潭、自己在小碧潭
    else {
        stop_count = Math.abs(friend - current);
    }

    // 計算站數
    // let stop_count = green_line_stations.indexOf(start) - green_line_stations.indexOf(end);

    return stop_count;
}

function findAndPrint(messages, currentStation){
    // your code here
    // 找到自己位置的捷運站 index
    let current_index = green_line_stations.indexOf(currentStation);


    let distance = friends_index.map( e => calulateStops(e,current_index )) ;
    // console.log(distance);
    
    // console.log(Math.min(...distance));
    // 朋友所在站的index
    // console.log(friends_index[distance.indexOf(Math.min(...distance))]);
    // console.log("the nearst is:")
    console.log(friends_location[friends_index[distance.indexOf(Math.min(...distance))]]);
    // console.log(green_line_stations[friends_index[distance.indexOf(Math.min(...distance))]]);
    // console.log("stop count: "+Math.min(...distance));
    
}




console.log("=== task1 ===");

findAndPrint(messages, "Wanlong"); // print Mary
findAndPrint(messages, "Songshan"); // print Copper
findAndPrint(messages, "Qizhang"); // print Leslie
findAndPrint(messages, "Ximen"); // print Bob
findAndPrint(messages, "Xindian City Hall"); // print Vivian

// task2

let time_table = Array.from({length: 24}, (_, i) => ["John","Bob","Jenny"]);

// console.log(time_table);
// console.log(time_table[15]);

function compare(available_consultants, criteria){

    if(available_consultants==""){
        return "no service";
    }
    if(criteria=="price"){
        // return lowest price from available_consultants
        return available_consultants.reduce((a, b ) => a.price < b.price ? a : b).name;

    }else if(criteria=="rate"){
        // return highest rate from available_consultants
        return available_consultants.reduce((a, b ) => a.rate > b.rate ? a : b).name;
    }else{
        return "exception";
    }
}

function book(consultants, hour, duration, criteria){
    // your code here
    
    // get who is available with the hour/duration
    // 取出各個時段有空的顧問
    let available_consultants = [];
    for (let index = hour; index < hour+duration; index++) {
        available_consultants.push(time_table[index]);
    }
    // console.log(available_consultants);

    // 取交集,得到可預約的顧問名單
    available_consultants_list = available_consultants.reduce((a, b) => a.filter(c => b.includes(c)));
    // console.log("return available consultant name:");
    // console.log("可預約的顧問名單");
    // console.log(available_consultants_list);

    // var options = ["a", "b"];
    // var results = ["c", "b", "a"];
    // var filtered = results.filter(result => options.indexOf(result) !== -1);

    let available_consultants_new = consultants.filter(consultant => available_consultants_list.indexOf(consultant.name) !== -1 )
    
    // return consultants based on criteria
    let recommendation = compare(available_consultants_new, criteria);
    // console.log("建議顧問:");
    console.log(recommendation);
    
    // update time_table
    for (let index = hour; index < hour+duration; index++) {
        let to_delete_index = time_table[index].indexOf(recommendation);
        if(to_delete_index!=-1){
            time_table[index].splice(to_delete_index,1);
        }
        // console.log(time_table[index]);
    }

    // print updated time_table
    // console.log(time_table);

}
const consultants=[
    {"name":"John", "rate":4.5, "price":1000},
    {"name":"Bob", "rate":3, "price":1200},
    {"name":"Jenny", "rate":3.8, "price":800}
];


console.log("=== task2 ===");
book(consultants, 15, 1, "price"); // Jenny
book(consultants, 11, 2, "price"); // Jenny
book(consultants, 10, 2, "price"); // John
book(consultants, 20, 2, "rate"); // John
book(consultants, 11, 1, "rate"); // Bob
book(consultants, 11, 2, "rate"); // No Service
book(consultants, 14, 3, "price"); // John


// task3
function getMiddleName(name){
    let len = name.length;
    switch(len){
        case 2:
            return name[1];
        case 4:
            return name[2];
        default:
            return name[Math.floor(name.length/2)];
    }

}
function func(...data){
    // ...data的意思
    // The rest parameter (...) allows a function to treat an indefinite number of arguments as an array:
    let str = [];
    let count = [];
    for (let arg of data) {
        let middle = getMiddleName(arg);
        // console.log(middle);
        if(str.indexOf(middle)==-1){
            str.push(middle);
            count[str.indexOf(middle)]=1;
            // console.log(str);
            // console.log(count);
        }else{
            str.push(middle);
            count.push(-1); // 表示該位置前面已經加總過了
            count[str.indexOf(middle)]++;
            // console.log(str);
            // console.log(count);
        }
    }
    // console.log(str);
    // console.log(count);
    // unique: 找出count陣列中只出現過一次的index
    let unique = count.indexOf(1);
    
    if(unique!=-1){
        console.log(data[unique]);
    }else{
        console.log("沒有");
    }
    
}
console.log("=== task3 ===");
func("彭大牆", "陳王明雅", "吳明"); // print 彭大牆
func("郭靜雅", "王立強", "郭林靜宜", "郭立恆", "林花花"); // print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花"); // print 沒有
func("郭宣雅", "夏曼藍波安", "郭宣恆"); // print 夏曼藍波安
// func("郭宣雅", "夏曼藍波安"); // print 夏曼藍波安

// task4

// 0,4,8,7,11,15,14,18,22,21,25...
// 取3餘數:
// 餘0: 前項-1
// 餘1: 前項+4
// 餘2: 前項+4

function getSequence(index){
    // 首項為0
    if(index == 0){
        return 0;
    }
    // 例外狀況
    else if(index<0){
        return false;
    }
    var idx = index%3;
    switch (idx) {
        case 1:
            // 餘1:前項+4
            return getSequence(index-1)+4;
            
        case 2: 
            // 餘2:前項+4
            return getSequence(index-1)+4;
            

        default:
            // 餘0:前項-1
            return getSequence(index-1)-1;
            
    }
}


function getNumber(index){
    console.log(getSequence(index));
}

console.log("=== task4 ===");
getNumber(1); // print 4
getNumber(5); // print 15
getNumber(10); // print 25
getNumber(30); // print 70


// task5

let return_available_index = (stat) => {
    let available_car = [];
    for (let index = 0; index < stat.length; index++) {
        if(stat[index]==1) available_car.push(index);
    }
    return available_car;
}

function find(spaces, stat, n){
    // your code here
    

    for (let index = 0; index < stat.length; index++) {
        if(stat[index]==0) spaces[index]=-1;
    }
    // console.log(spaces);
    if(Math.max(...spaces) <n){console.log(-1);}
    else{
        // let smallest_i = 0;
        // for (let i = 0; i < spaces.length; i++) {
            
        //     if(spaces[i]>=n){
        //         if(spaces[smallest_i]>=spaces[i]){ smallest_i = i;}
        //     }else{
        //         smallest_i++;
        //     }   
            
        // }
        // let x = spaces.filter(e => e>=n);
        
        // console.log( spaces.indexOf(Math.min(...x)) );
        console.log(spaces.indexOf(Math.min(...spaces.filter(e => e>=n))));

        // console.log(spaces.findIndex(function(e){
        //     Math.min(e);
        // }));

        // console.log(smallest_i);
    }
    
    
}

console.log("=== task5 ===");
find([3, 1, 5, 4, 3, 2], [0, 1, 0, 1, 1, 1], 2); // print 5
find([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4); // print -1
find([4, 6, 5, 8], [0, 1, 1, 1], 4); // print 2

File: week2/task.py
# task1

# 原始訊息的dict
messages={
    "Leslie":"I'm at home near Xiaobitan station.",
    "Bob":"I'm at Ximen MRT station.",
    "Mary":"I have a drink near Jingmei MRT station.",
    "Copper":"I just saw a concert at Taipei Arena.",
    "Vivian":"I'm at Xindian station waiting for you."
}
# 捷運站名的list
green_line_stations = ["Songshan","Nanjing Sanmin","Taipei Arena","Nanjing Fuxing","Songjiang Nanjing","Zhongshan","Beimen","Ximen","Xiaonanmen","Chiang Kai-Shek Memorial Hall","Guting","Taipower Building","Gongguan","Wanlong","Jingmei","Dapinglin","Qizhang","Xindian City Hall","Xindian","Xiaobitan"];

# 找出朋友所在的index
# 把message中有比對到跟站名list一樣的字串找出來
# 並製作 朋友名字: 站名: 捷運站index: 的dict
friend_station_list = [ {"name":name,"station":i,"station_idx":green_line_stations.index(i)} for i in green_line_stations for name in messages if messages[name].find(i) > -1 ]
# print(friend_station_list)
# [{'name': 'Copper', 'station': 'Taipei Arena', 'station_idx': 2}, 
#  {'name': 'Bob', 'station': 'Ximen', 'station_idx': 7}, 
#  {'name': 'Mary', 'station': 'Jingmei', 'station_idx': 14}, 
#  {'name': 'Vivian', 'station': 'Xindian', 'station_idx': 18}, 
#  {'name': 'Leslie', 'station': 'Xiaobitan', 'station_idx': 19}]

########## 下面3行簡化成上面1行 ##########
# friend_idx = [ i for i in green_line_stations for name in messages if messages[name].find(i) > -1 ]
# friend_station = [ name for i in green_line_stations for name in messages if messages[name].find(i) > -1 ]
# friend_station_index = [ green_line_stations.index(key) for key in friend_idx  ]
########################################

# 取出站名對應idx
friend_station_index = [ i['station_idx'] for i in friend_station_list]
# print(friend_station_index) # ex: [2, 7, 14, 18, 19]

# 取出朋友名字列表
friend_station = [ i['name'] for i in friend_station_list]
# print(friend_station) # ex: ['Copper', 'Bob', 'Mary', 'Vivian', 'Leslie']

# 計算相距站數
def calculate_stops(friend, current):
    # 狀況1: 朋友在小碧潭、自己不在小碧潭: 計算從自己到七張的站數,再+1
    # Friend at Xiaobitan, current not at Xiaobitan
    if friend == 19 and current != 19:
        # Calculate the number of stations from Qizhang to current location + 1
        stop_count = abs(16 - current) + 1
    
    # 狀況2: 朋友不在小碧潭、自己在小碧潭: 計算從朋友到七張的站數,再+1
    # Friend not at Xiaobitan, current at Xiaobitan
    elif friend != 19 and current == 19:
        stop_count = abs(friend - 16) + 1
    
    # 其他狀況: 朋友跟自己都不在小碧潭、or 朋友跟自己都在小碧潭: 計算從朋友到自己的站數
    # Friend not at Xiaobitan, current not at Xiaobitan OR Friend at Xiaobitan, current at Xiaobitan
    else:
        stop_count = abs(friend - current)
    
    # 回傳站數
    return stop_count

def find_and_print(messages, current_station):
    # your code here
    # 取出自己目前所在位置的捷運站index
    current_index = green_line_stations.index(current_station)

    # 計算每個朋友到自己所在位置的站數,存成list
    distance_list = [ calculate_stops(i, current_index) for i in friend_station_index ]
    # print(distance_list) # ex: [11, 6, 1, 5, 4]

    # 找出最少站數的index 並對應到朋友名字列表, 得出距離最近的朋友
    print(friend_station[distance_list.index(min(distance_list))])
    

print("=== task1 ===")
find_and_print(messages, "Wanlong") # print Mary
find_and_print(messages, "Songshan") # print Copper
find_and_print(messages, "Qizhang") # print Leslie
find_and_print(messages, "Ximen") # print Bob
find_and_print(messages, "Xindian City Hall") # print Vivian

# task2


consultants=[
    {"name":"John", "rate":4.5, "price":1000},
    {"name":"Bob", "rate":3, "price":1200},
    {"name":"Jenny", "rate":3.8, "price":800}
]

# 建立一個初始時間表(list), 有24個element(24小時)每個element都有三個顧問的名字，表示每個顧問都可以預約
# time_table = [ consultants for i in range(24) ]
# 修正: consultants資料會變動, 所以不能寫死
time_table = [ [ i['name'] for i in consultants ] for i in range(24) ]
# print(time_table)



# 傳入有空的顧問list,以及比對的標準(criteria)
def compare(available_consultants_name, criteria):
    # print(available_consultants_name) # ex: ["John","Bob"]

    # 從顧問的dict中取出有空的顧問,以及其 rating/price 資料
    candidates = [ i for i in consultants for j in available_consultants_name if i['name'].find(j)!=-1  ]
    # print(candidates) # ex: [{'name': 'John', 'rate': 4.5, 'price': 1000}, {'name': 'Bob', 'rate': 3, 'price': 1200}]

    # 沒有顧問有空時: 無法服務
    if(len(available_consultants_name)==0):
        return "no service"
    
    if(criteria=="rate"):
        # 從candidates中取出有空顧問的rating
        rate_list = [ i['rate'] for i in candidates ]
        # 從candidates中取出有有最高rating的顧問
        max_rate_name = [ i['name'] for i in candidates if i['rate'] == max(rate_list) ]
        # 回傳顧問名字
        return max_rate_name[0]
     
    if(criteria=="price"):
        # 從candidates中取出有空顧問的price
        price_list = [ i['price'] for i in candidates ]
        # 從candidates中取出有有最低price的顧問
        min_price_name = [ i['name'] for i in candidates if i['price'] == min(price_list) ]
        # 回傳顧問名字
        return min_price_name[0]

def book(consultants, hour, duration, criteria):
    
    # // get who is available with the hour/duration
    # // 從預約表(time_table)中取出指定時段(hour/duration)有空的顧問
    available_consultants = []
    for i in range(hour, hour+duration):
        available_consultants.append(time_table[i])
    # print(available_consultants) # ex: [['John', 'Bob', 'Jenny'], ['John', 'Bob']]

    # // 取交集,得到可預約的顧問名單
    available_consultants_set = set(available_consultants[0]) # 型別轉換,把list改成set,不然無法取交集
    for i in available_consultants:
        available_consultants_set &= set(i) # 取交集
    # print(available_consultants_set) # {'Bob', 'John'}
    # print(list(available_consultants_set)) # 如果想改回list: ['Bob', 'John']

    # 根據有空的名單以及條件找出推薦顧問
    recommendation = compare(available_consultants_set, criteria)
    print(recommendation)

    # 更新預約表 把推薦的顧問從預約的時段中刪除
    # update time_table
    for i in range(hour, hour+duration):
        if(recommendation!="no service"):
            time_table[i].remove(recommendation)
    


print("=== task2 ===")
book(consultants, 15, 1, "price") # Jenny
book(consultants, 11, 2, "price") # Jenny
book(consultants, 10, 2, "price") # John
book(consultants, 20, 2, "rate") # John
book(consultants, 11, 1, "rate") # Bob
book(consultants, 11, 2, "rate") # No Service
book(consultants, 14, 3, "price") # John


# task3
# 取中間字的方法
def getMiddleName(name):
    length = len(name)
    match length:
        case 2:
            return name[1]
        case 4:
            return name[2]
        case _: # 其他狀況: 除2無條件捨去
            return name[length//2]

def func(*data):
    str = []
    count = []
    for x in data:
        middle = getMiddleName(x)
        # 判斷取出來的字是不是已經存在str中了,如果是則設為該名字list的index ;否則設為-1
        middle_index = str.index(middle) if middle in str else -1
        
        # 沒有存過, 把取出來的字存到str中, 並在count list中放1
        if(middle_index==-1):
            str.append(middle)
            count.append(1)
        # 有存過, 把取出來的字存到str中, 並新增一個-1, 表示該位置不需要計算總數,並把前面對應的count加1
        else:
            # print("已有存過middle")
            # print(middle_index)
            str.append(middle)
            count.append(-1) # 表示該位置前面已經加總過了
            count[middle_index] += 1
    # print(str) # ex: ['靜', '立', '靜', '立', '花']
    # print(count) # ex: [2, 2, -1, -1, 1]
    
    # python 的index方法 如果沒找到會出現error 不會回傳-1 (跟javascript不一樣)
    # 解決辦法:
    # unique: 找count為1的index; 如果找不到則設為-1
    # 注意: 這裡會有一個情況: 如果count中有多個出現一次的中間名,則只會回傳第一個找到的
    unique = count.index(1) if 1 in count else -1
    unique_middlename = data[unique] if(unique!=-1) else "沒有" 
    print(unique_middlename)


print("=== task3 ===")
func("彭大牆", "陳王明雅", "吳明") # print 彭大牆
func("郭靜雅", "王立強", "郭林靜宜", "郭立恆", "林花花") # print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花") # print 沒有
func("郭宣雅", "夏曼藍波安", "郭宣恆") # print 夏曼藍波安
# // func("郭宣雅", "夏曼藍波安"); // print 夏曼藍波安


# task4

# 0,4,8,7,11,15,14,18,22,21,25....
# 取3餘數:
# 餘0: 前項-1
# 餘1: 前項+4
# 餘2: 前項+4

def get_sequence(index):
    if(index == 0):
        return 0
    idx = index % 3
    if(idx == 1):
        return get_sequence(index-1)+4
    elif(idx == 2):
        return get_sequence(index-1)+4
    return get_sequence(index-1)-1

def get_number(index):
    print(get_sequence(index))

print("=== task4 ===")
get_number(1) # print 4
get_number(5) # print 15
get_number(10) # print 25
get_number(30) # print 70


# task5

def find(spaces, stat, n):
    # your code here
    
    # stat為0表示該車廂不開放, 把相對應的spaces list 設為-1
    for index, item in enumerate(stat):
        if(stat[index]==0):
            spaces[index] = -1
    
    # 如果spaces中所有位置總數都比需求(n)小, 表示無車可搭 回傳-1
    if(max(spaces) < n):
        print(-1)
    # 找出比n大的space中,最小的那個車廂Index並回傳
    else:
        smallest_i = 0
        for index, item in enumerate(spaces):
            if(spaces[index]>=n):
                if(spaces[smallest_i]>=spaces[index]):
                    smallest_i = index
            else:
                smallest_i += 1
        print(spaces.index(min([x for x in spaces if x>=n])))

print("=== task5 ===")
find([3, 1, 5, 4, 3, 2], [0, 1, 0, 1, 1, 1], 2) # print 5
find([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4) # print -1
find([4, 6, 5, 8], [0, 1, 1, 1], 4) # print 2
